core/src/main/java/org/apache/accumulo/core/iterators/ClientIteratorEnvironment.java

```diff
@@ -0,0 +1,220 @@
+    private Optional<SamplerConfiguration> samplerConfig = Optional.empty();
+    private boolean samplingEnabled = false;
+    protected Optional<ClientServiceEnvironmentImpl> env = Optional.empty();
+
+    public Builder withScope(IteratorScope scope) {
+      this.scope = Optional.of(scope);
+      return this;
+    }
+
+    public Builder isFullMajorCompaction() {
+      this.isFullMajorCompaction = true;
+      return this;
+    }
+
+    public Builder withAuthorizations(Authorizations auths) {
+      this.auths = Optional.of(auths);
+      return this;
+    }
+
+    public Builder isUserCompaction() {
+      this.isUserCompaction = true;
+      return this;
+    }
+
+    public Builder withTableId(TableId tableId) {
+      this.tableId = Optional.of(tableId);
+      return this;
+    }
+
+    public Builder withSamplingEnabled() {
+      this.samplingEnabled = true;
+      return this;
+    }
+
+    public Builder withSamplerConfiguration(SamplerConfiguration sc) {
+      this.samplerConfig = Optional.ofNullable(sc);
+      return this;
+    }
+
+    public Builder withClient(AccumuloClient client) {
+      this.env = Optional.of(new ClientServiceEnvironmentImpl((ClientContext) client));
+      return this;
+    }
+
+    public ClientIteratorEnvironment build() {
+      return new ClientIteratorEnvironment(this);
+    }
+
+  }
+
+  public static final IteratorEnvironment DEFAULT = new Builder().build();
+
+  private final Optional<IteratorScope> scope;
+  private final boolean isFullMajorCompaction;
+  private final Optional<Authorizations> auths;
+  private final boolean isUserCompaction;
+  private final Optional<TableId> tableId;
+  private final Optional<SamplerConfiguration> samplerConfig;
+  private final boolean samplingEnabled;
+  private final Optional<ClientServiceEnvironmentImpl> env;
+
+  private ClientIteratorEnvironment(Builder builder) {
+    this.scope = builder.scope;
+    this.isFullMajorCompaction = builder.isFullMajorCompaction;
+    this.auths = builder.auths;
+    this.isUserCompaction = builder.isUserCompaction;
+    this.tableId = builder.tableId;
+    this.samplerConfig = builder.samplerConfig;
+    this.env = builder.env;
+    this.samplingEnabled = builder.samplingEnabled;
+  }
+
+  /**
+   * Copy constructor used for enabling sample. Only called from {@link cloneWithSamplingEnabled}.
+   */
+  private ClientIteratorEnvironment(ClientIteratorEnvironment copy) {
+    this.scope = copy.scope;
+    this.isFullMajorCompaction = copy.isFullMajorCompaction;
+    this.auths = copy.auths;
+    this.isUserCompaction = copy.isUserCompaction;
+    this.tableId = copy.tableId;
+    this.samplerConfig = copy.samplerConfig;
+    this.env = copy.env;
+    this.samplingEnabled = true;
+  }
+
+  @Override
+  @Deprecated(since = "2.0.0")
+  public SortedKeyValueIterator<Key,Value> reserveMapFileReader(String mapFileName)
+      throws IOException {
+    Configuration hadoopConf = new Configuration();
```

Comment:
Was this implementation based on something else?  On the server side this method can open an rfile or a mapfile (Looking at TabletIteratorEnvironment).  According to my IDE nothing calls this method, so no test.  Since this method has zero test, we may be better off throwing UnsupportedOperationException for now.  Should probably test it if we add an impl.

Reply:
It came from core/test/.../DefaultIteratorEnvironment. I also found no references to this method, but wasn't sure if we should try and implement something, or throw an exception. I'm good with changing this to throwing an exception.

Reply:
Modified this in 2d39e46 to throw an UnsupportedOperationException